flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/FunctionDefinitionFactory.java

```diff
@@ -34,4 +35,29 @@ public interface FunctionDefinitionFactory {
      * @return a {@link FunctionDefinition}
      */
     FunctionDefinition createFunctionDefinition(String name, CatalogFunction catalogFunction);
+
+    /**
+     * Creates a {@link FunctionDefinition} from given {@link CatalogFunction}. If the {@link
+     * CatalogFunction} is created by user defined resource, the user of {@link
+     * FunctionDefinitionFactory} needs to override this method explicitly. The implementation logic
+     * needs to use the user classloader to load custom classes instead of the thread context
+     * classloader.
+     *
+     * @param name name of the {@link CatalogFunction}
+     * @param catalogFunction the catalog function
+     * @param userClassLoader the class loader is used to load user defined function's class
+     * @return a {@link FunctionDefinition}
+     */
+    default FunctionDefinition createFunctionDefinition(
+            String name, CatalogFunction catalogFunction, ClassLoader userClassLoader) {
```

Comment:
Intead of providing ClassLoader directly, how about providing a `Context` which contains `ClassLoader` just like `ModuleFactory` `DynamicTableFactory` does?
So that, if we need other info for this method in the future, we won't need to modify function signature.

Reply:
We can't get the `Context` in `FunctionCatalog`, so this maybe not possible. BTW, there are different `Context` interface for `ModuleFactory`, `DynamicTableFactory` etc..., So I think this is unfeasible. If you need `ModuleManager`, we can add it function now.

Reply:
IMO, `FunctionDefinitionFactory` is only used to hive, do you think here wether exist the possible to use `HiveModule`? I have revert these changes, you help provide the implementation [FLINK-28451](https://issues.apache.org/jira/browse/FLINK-28451) according to your needs? Please also take a look about FLIP-214.

Reply:
Ok, make sense to me.